### PR TITLE
feat(pageserver): generate image layers for sparse keyspace

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

Part of https://github.com/neondatabase/neon/issues/7462

Sparse keyspace does not generate image layers for now. This pull request adds support for generating image layers for sparse keyspace.

## Summary of changes

* Use the scan interface to generate compaction data for sparse keyspace.
* Track num of delta layers reads during scan.
* Read-trigger compaction: when a scan on the keyspace touches too many delta files, generate an image layer. There are one hard-coded threshold for now: max delta layers we want to touch for a scan.
* L0 compaction does not need to compute holes for metadata keyspace.

Know issue: the scan interface currently reads past the image layer, which causes `delta_layer_accessed` keeps increasing even if image layers are generated. The pull request to fix that will be separate, and orthogonal to this one.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by skyzh** [comment]
For reference, this is the log of test_metadata_scan, where image layer generation happens and L0 compaction works correctly.

[test_metadata_scan.log](https://github.com/neondatabase/neon/files/15312878/test_metadata_scan.log)

File: pageserver/src/tenant.rs
Comment (VladLazar) [inline pageserver/src/tenant.rs:5621]: Can you add a check that compaction actually did what you expect?

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:4111]: accidental change?

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:4240]: Can you separate the metadata and non-metadata logic into different functions?

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:4328]: I find it hard to reason about this. What does this context track? It's specific for compaction, so it should be all reads done during compaction, but what are those?

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:4354]: Please lift these into named constants.

Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:4354]: I'm not a big fan on tracking how many bytes were read in each request. It's done for every request, but we don't need it every time as far as I understand. Could you use the number of layers visited instead?

File: pageserver/src/tenant.rs
Comment (skyzh) [reply pageserver/src/tenant.rs:5621]: If it doesn't crash + runs in a reasonable amount of time, it works, so I don't plan to add checks for now.

Once the vectored get interface can stop when an image layer covers all keyspace, I will add a check that `vectored_access_delta_file_cnt` is decreasing for each round of compaction/force_image_layer_creation.

File: pageserver/src/tenant/timeline.rs
Comment (skyzh) [reply pageserver/src/tenant/timeline.rs:4354]: yep, I'll change it to num delta layers visited for now, and find better ways to model it. The reason why I used kbs instead of num of files is that it's hard to compute read amplification without the size information. But I think I can make a separate PR with benchmark/testing result for further discussion.

File: pageserver/src/tenant/timeline.rs
Comment (skyzh) [reply pageserver/src/tenant/timeline.rs:4328]: All reads done during compaction. We use <number after compaction> - <number before compaction> to estimate the number of kbs read in the process.

File: pageserver/src/tenant/timeline.rs
Comment (skyzh) [reply pageserver/src/tenant/timeline.rs:4405]: https://github.com/neondatabase/neon/pull/7567/commits/5af19aa627fd269875e52a40ac52f0d76fb22bc8

File: pageserver/src/tenant/timeline.rs
Comment (skyzh) [reply pageserver/src/tenant/timeline.rs]: https://github.com/neondatabase/neon/pull/7567/commits/5af19aa627fd269875e52a40ac52f0d76fb22bc8

File: pageserver/src/tenant.rs
Comment (VladLazar) [reply pageserver/src/tenant.rs:5621]: > If it doesn't crash + runs in a reasonable amount of time + does not lose any data, it works, so I don't plan to add checks for now.

It would also "work" if no image layers are added.

File: pageserver/src/tenant/timeline.rs
Comment (arpad-m) [inline pageserver/src/tenant/timeline.rs:4164]: it's only creating up to one image layer here, right? it is called `create_image_layers_for_rel_blocks` however.

The returned Vec has up to one item, so maybe return an option instead? the `extend` function should support this as well.

File: pageserver/src/tenant/timeline.rs
Comment (arpad-m) [inline pageserver/src/tenant/timeline.rs:4259]: same here: we create either a single image layer, or nothing.

File: pageserver/src/tenant/timeline.rs
Comment (arpad-m) [inline pageserver/src/tenant/timeline.rs:4172]: as discussed in the call: somewhere in this function we should also include an error (either logged or aborting compaction entirely) if the written image layer is larger than the target size. if we get to the upload limit, we should bail out of compaction entirely (until partitioning is implemented).

File: pageserver/src/tenant/timeline.rs
Comment (skyzh) [reply pageserver/src/tenant/timeline.rs:4261]: fixed in https://github.com/neondatabase/neon/pull/7567/commits/0823f86e4408d22a3764b6fd41d2780a1076e9d1

File: pageserver/src/tenant.rs
Comment (skyzh) [reply pageserver/src/tenant.rs:5621]: added a test case on l0 compaction check https://github.com/neondatabase/neon/pull/7567/commits/6eaacbf14d8fc3e5c18239296c46d0324fec7f56

**Comment by skyzh** [comment]
@arpad-m @VladLazar comments resolved and ready for review, thanks :)

File: pageserver/src/tenant/storage_layer.rs
Comment (VladLazar) [inline pageserver/src/tenant/storage_layer.rs:154]: Once you merge https://github.com/neondatabase/neon/pull/7773, `get_layers_visited` will become a very good approximation for this. Since you're not splitting image layers, it will basically return the number of delta layers plus one. Hence, I don't think you need this and the associated change in `on_layers_visited`.

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:1218]: This is a hot path. Doesn't feel right to add an atomic increment here which only compaction cares about.

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:4389]: did you mean to add the `;`?

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:4261]: nit: `create_image_layer_for_metadata_keys`

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:4282]: Why not `Timeline::scan`? At least that does some throttling...

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:4279]: You keep this atomic in the request context which is really only needed here.
Here's an idea: Make `get_vectored_impl` take a mutable ref to `ValuesReconstructState`, lift the `ValuesReconstructState` up and inspect the number of layers visited on that to figure out if you should add an image.

```
let mut reconstruct_state = ValueReconstructState::default();
let data = self.get_vectored_impl(..., &mut reconstruct_state, ...);
let layers_visited = reconstruct_state.get_layers_visited();
... rest of your logic here ...
```

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:4301]: nit: can you split this line?

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:4316]: Again, I'd pre-emptively create an issue for this.

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:4271]: nit: `assert!(!matches!(mode, ImageLayerCreationMode::Initial))` would save from having to jump back to the caller.

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:4311]: I don't understand what situation you're referring to. If this is a bug, please create an issue and link it here.

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:4166]: nit: `create_image_layer_for_rel_blocks`

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline.rs:4173]: Can we use a return value instead? The function can return something like `Result<ImageLayerCreationOutcome>` where `ImageLayerCreationOutcome` is:
```
struct ImageLayerCompactionOutcome {
  image: Option<ResidentLayer>,
  compacted_up_to: Key
};
```

And then you can handle the `start` update in the same function as it was declared.

File: pageserver/src/tenant/timeline/compaction.rs
Comment (VladLazar) [inline pageserver/src/tenant/timeline/compaction.rs:496]: Not sure I understand what's going on here. Do you have an example?

**Review by VladLazar — COMMENTED** [review]
I'd really like to avoid adding something to the request context that is only needed for compaction. I've got one suggestion here.

File: pageserver/src/tenant/timeline/compaction.rs
Comment (skyzh) [reply pageserver/src/tenant/timeline/compaction.rs:496]: For metadata keys the keyspace is always sparse, and the algorithm will try to create too many holes / partitions. We just skip doing partition for metadata keys here so that it's always handled as a whole in the compaction and image layer creations.

File: pageserver/src/tenant/timeline.rs
Comment (skyzh) [reply pageserver/src/tenant/timeline.rs:4311]: I just don't feel safe of directly removing a single key from the image layer because we didn't do this before. https://github.com/neondatabase/neon/pull/7773 might fix that, but I need to create a test case that ensures everything works before uncommenting this.

File: pageserver/src/tenant/timeline.rs
Comment (skyzh) [reply pageserver/src/tenant/timeline.rs:4316]: https://github.com/neondatabase/neon/issues/7796

File: pageserver/src/tenant/storage_layer.rs
Comment (skyzh) [reply pageserver/src/tenant/storage_layer.rs:154]: but it's still an approximation, and once someone modifies the image layer generation logic, they will likely spend some time fixing existing test cases relying on that assumption...? So I'm in favor of adding this statistics, and as you said, in a more light-weight way inside reconstruct value state.

File: pageserver/src/tenant/timeline.rs
Comment (skyzh) [reply pageserver/src/tenant/timeline.rs:4282]: I thought compaction (or create image layer) itself is already throttled and we don't need to throttle it again?

File: pageserver/src/tenant/storage_layer.rs
Comment (VladLazar) [reply pageserver/src/tenant/storage_layer.rs:154]: > but it's still an approximation, and once someone modifies the image layer generation logic, they will likely spend some time fixing existing test cases relying on that assumption...?

Not sure what you mean.

> in a more light-weight way inside reconstruct value state

+1

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [reply pageserver/src/tenant/timeline.rs:4282]: It's throttled in the sense that we limit the number of concurrent compactions that run at once. Under the assumption that metadata key space is very sparse not throttling is not a big deal, but then again compaction for rel blocks gets throttled correctly.

Comment (VladLazar) [reply pageserver/src/tenant/timeline.rs:4282]: Anyways, the difference between one semaphore unit and 0 semaphore units is kinda insignificant

Comment (skyzh) [reply pageserver/src/tenant/timeline.rs]: Another reason I want to keep using `get_vectored_impl` is that we need to get `ValuesReconstructState` back... where `scan` does not expose this to the user of the interface.

File: pageserver/src/tenant/timeline.rs
Comment (skyzh) [reply pageserver/src/tenant/timeline.rs:4307]: https://github.com/neondatabase/neon/issues/7800

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [reply pageserver/src/tenant/timeline.rs]: Yeah, my comments were contradictory in that regard. Leaving it up to you to decide if we should worry about throttling for now

File: pageserver/src/tenant/storage_layer.rs
Comment (skyzh) [reply pageserver/src/tenant/storage_layer.rs:158]: ... also need to consider if we generate multiple image layers for the metadata keyspace (which I think will happen very soon), so I think it's good to have these two separate metrics on layer visited for now.

File: pageserver/src/tenant/timeline.rs
Comment (skyzh) [reply pageserver/src/tenant/timeline.rs]: renamed compacted_up_to -> next_start_key because create image layer is not actually a compaction.

**Comment by skyzh** [comment]
should be ready for review and wait for CI runs...

File: pageserver/src/tenant/storage_layer.rs
Comment (VladLazar) [reply pageserver/src/tenant/storage_layer.rs:158]: > also need to consider if we generate multiple image layers for the metadata keyspace

That will break the scan interface with the changes in #7773.

In any case `layers_visited` would be off by at most the number of image layers, but if you feel strongly about it, then let's keep it

**Review by VladLazar — APPROVED** [review]
Looks good to me.

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [reply pageserver/src/tenant/timeline.rs]: I think it's fine to address Arpad's concern as a follow up since aux v2 is not enabled fleet wide.

**Pull request merged**
